libp2p/peerstore.nim

```diff
@@ -31,35 +30,59 @@ type
   # Each book contains a book (map) and event handler(s)
   AddressBook* = object
     book*: Table[PeerID, HashSet[MultiAddress]]
-    addrChange: AddrChangeHandler
+    changeHandlers: seq[AddrChangeHandler]
+  
+  ProtoBook* = object
+    book*: Table[PeerID, HashSet[string]]
+    changeHandlers: seq[ProtoChangeHandler]
+  
+  KeyBook* = object
+    book*: Table[PeerID, PublicKey]
+    changeHandlers: seq[KeyChangeHandler]
+  
+  MetadataBook* = object
+    book*: Table[PeerID, Table[string, seq[byte]]]
+    changeHandlers: seq[MetadataChangeHandler]
   
   ####################
   # Peer store types #
   ####################
 
   PeerStore* = ref object of RootObj
     addressBook*: AddressBook
-    listeners: seq[EventListener]
+    protoBook*: ProtoBook
+    keyBook*: KeyBook
+    metadataBook*: MetadataBook
   
   StoredInfo* = object
     # Collates stored info about a peer
-    ## @TODO include data from other stores once added
     peerId*: PeerID
     addrs*: HashSet[MultiAddress]
+    protos*: HashSet[string]
+    publicKey*: PublicKey
+    metadata*: Table[string, seq[byte]]
 
-proc init(T: type AddressBook, addrChange: AddrChangeHandler): AddressBook =
+proc init(T: type AddressBook): AddressBook =
   T(book: initTable[PeerId, HashSet[MultiAddress]](),
-    addrChange: addrChange)
+    changeHandlers: newSeq[AddrChangeHandler]())
```

Comment:
Just a small remark - there is no need to initialize the sequence explicitly with `newSeq[T]()`, `seq[T]` does auto initialization, just like a `string` does.

Reply:
Thanks! And addressed: 926d340